lib/vm/src/instance/mod.rs

```diff
@@ -113,6 +107,139 @@ pub(crate) struct Instance {
     vmctx: VMContext,
 }
 
+/// A collection of data about host envs.
+///
+/// Due to invariants in the `HostEnv` variant, this type should only
+/// be constructed by one of the given methods, not directly.
+#[derive(Debug)]
+pub enum ImportFunctionEnv {
+    /// The `vmctx` pointer does not refer to a host env, there is no
+    /// metadata about it.
+    NoEnv,
```

Comment:
I think is better to name this: `VMEnv`.
And have the `HostEnv` to be an enum with all the different options:
* NoEnv
* Env

Also, the VM should not be aware of dynamic functions or not. So I believe `DynamicHostEnvWithNoInnerEnv` should be done via `HostEnv`

Reply:
Same thought on the `VM` prefix, I'm not sure it makes sense to use it on types that aren't related to the VM.

Yeah, the extra case here is just to make the multiple cases more explicit.  Otherwise there's data that's sometimes there and sometimes not and it's not exactly clear why. This enum is more descriptive than prescriptive: it's describing the state of the world as it is, not what it should be. Perhaps we should move this type out of `vm`, there's no reason it has to be defined in VM.

Reply:
For the VM, there is not such thing as DynamicHostEnv. This is something that we do in the engine level, not VM

Reply:
Meaning: the `vm` only handles the functions with their end ABI. Is the engine the one that creates the trampolines and so.

Reply:
Note: there is some "dangling" code in the VM that is aware of dynamic functions. But thanks to the PR where we move most of the function stuff to the engine, it should be easy to move the rest as well.